File: apps/api/engine/checks/tool_poisoning.py
import re
from ..models import Finding, Severity, OWASPCategory
from ..parser import MCPServer

_INJECTION_PATTERNS: list[tuple[re.Pattern, str]] = [
    (re.compile(r'ignore\s+(previous|all|prior)\s+instructions?', re.I), "instruction override phrase"),
    (re.compile(r'do\s+not\s+(reveal|disclose|share|tell)', re.I), "information suppression directive"),
    (re.compile(r'(override|bypass|ignore)\s+(the\s+)?(system\s+prompt|instructions?|rules?|constraints?)', re.I),
     "system prompt override"),
    (re.compile(r'you\s+(are|must|should)\s+(now\s+|always\s+)?(act\s+as|pretend|roleplay|simulate)', re.I),
     "role manipulation directive"),
    (re.compile(r'(hidden|secret|confidential)\s+instructions?', re.I), "hidden instruction reference"),
    (re.compile(r'<\|.+?\|>', re.I), "control token injection"),
    # LLM control tokens: Llama [INST], Claude <<SYS>>, GPT-4 ###
    (re.compile(r'\[INST\]|\[/INST\]|<<SYS>>|<</SYS>>', re.I), "LLM control token"),
    (re.compile(r'###\s*SYSTEM\s*###|###\s*USER\s*###|###\s*ASSISTANT\s*###', re.I), "LLM section delimiter"),
    # XML-style system prompt tags used in Claude 3.5+ multi-turn injection
    (re.compile(r'<system>|</system>|<user>|</user>|<assistant>|</assistant>', re.I), "XML role delimiter injection"),
    (re.compile(r'AUTOEXECUTE|AUTO_EXECUTE|SYSTEM_OVERRIDE', re.I), "automation override keyword"),
    (re.compile(r'disregard\s+(all|any|previous)', re.I), "instruction disregard phrase"),
    (re.compile(r'(your|the)\s+(real|true|actual)\s+(purpose|goal|mission|task)\s+is', re.I), "purpose redefinition"),
    (re.compile(r'developer\s+mode|jailbreak\s+mode|DAN\s+mode', re.I), "jailbreak mode reference"),
]

# Data exfiltration indicators in args (all patterns — AS-017 equivalent)
_EXFILTRATION_PATTERNS: list[tuple[re.Pattern, str]] = [
    (re.compile(r'(?i)(send|upload|transmit|exfiltrat|forward|relay)\s+(data|files?|credentials?|tokens?)\s+(to|from)'), "data transfer directive"),
    (re.compile(r'(?i)POST\s+to\s+https?://'), "HTTP POST to external URL"),
    (re.compile(r'https?://(?!localhost|127\.0\.0\.1|0\.0\.0\.0)[^\s]{15,}'), "external URL embedded in argument"),
    (re.compile(r'(?i)(webhook|callback)\s+url'), "webhook/callback URL reference"),
    (re.compile(r'(?i)(steal|harvest|collect|scrape)\s+(data|credentials?|tokens?|passwords?)'), "data harvesting language"),
    (re.compile(r'(?i)(bcc|blind.?carbon.?copy)'), "BCC/blind copy email exfiltration"),
    (re.compile(r'(?i)(forward.?to|cc.?to|reply.?to)\s*[=:]\s*[^\s@]+@[^\s]{3,}'), "email forwarding rule"),
]

# Subset safe to apply to env var VALUES — excludes generic URL pattern (too many false positives
# from legitimate API endpoint env vars like BACKEND_URL=https://api.example.com).
_EXFILTRATION_ENV_PATTERNS: list[tuple[re.Pattern, str]] = [
    (re.compile(r'(?i)(send|upload|transmit|exfiltrat|forward|relay)\s+(data|files?|credentials?|tokens?)\s+(to|from)'), "data transfer directive"),
    (re.compile(r'(?i)POST\s+to\s+https?://'), "HTTP POST to external URL"),
    (re.compile(r'(?i)(webhook|callback)\s+url'), "webhook/callback URL reference"),
    (re.compile(r'(?i)(steal|harvest|collect|scrape)\s+(data|credentials?|tokens?|passwords?)'), "data harvesting language"),
    # Postmark Sep 2025: DEFAULT_BCC env var silently exfiltrated all emails to attacker for 2 weeks
    (re.compile(r'(?i)(bcc|blind.?carbon.?copy)'), "BCC/blind copy email exfiltration"),
    (re.compile(r'(?i)(forward.?to|cc.?to|reply.?to)\s*[=:]\s*[^\s@]+@[^\s]{3,}'), "email forwarding rule"),
]

_MAX_ARGS_LENGTH = 2000
_HORIZONTAL_SCROLL_THRESHOLD = 300  # chars on a single line before flagging PI-003

# PI-004: Obfuscation via escape sequences
# 4+ consecutive unicode escapes = "bash" or "ignore" hiding in plain sight
_UNICODE_ESC_RE = re.compile(r'(?:\\u[0-9a-fA-F]{4}){4,}')
# 4+ consecutive hex escapes = same technique in hex form
_HEX_ESC_RE = re.compile(r'(?:\\x[0-9a-fA-F]{2}){4,}')

# PI-005: Invisible / zero-width / bidi-override Unicode characters (Research 1).
# Distinct from PI-004: PI-004 catches \uXXXX as a literal 6-char text sequence.
# PI-005 catches the actual decoded codepoints already present in the config string.
# SH-004 catches non-ASCII *letters* (Unicode category "L") in server names.
# PI-005 targets category "Cf" (Format) chars in args/env — a vector SH-004 misses.
# Reference: Trojan Source CVE-2021-42574; MDPI May 2026 MCP Threat Modeling paper.
_INVISIBLE_UNICODE: dict[str, str] = {
    '​': 'Zero Width Space',
    '‌': 'Zero Width Non-Joiner',
    '‍': 'Zero Width Joiner',
    '﻿': 'BOM/Zero Width No-Break Space',
    '⁠': 'Word Joiner',
    '⁡': 'Function Application (invisible)',
    '⁢': 'Invisible Times',
    '⁣': 'Invisible Separator',
    '⁤': 'Invisible Plus',
    '­': 'Soft Hyphen',
    '᠎': 'Mongolian Vowel Separator',
    '͏': 'Combining Grapheme Joiner',
}
# Bidi overrides can flip text display direction — U+202E makes text read backwards in the UI.
_BIDI_OVERRIDE_UNICODE: dict[str, str] = {
    '‪': 'LTR Embedding',
    '‫': 'RTL Embedding',
    '‬': 'Pop Directional Formatting',
    '‭': 'LTR Override',
    '‮': 'RTL Override (Trojan Source)',
    '⁦': 'LTR Isolate',
    '⁧': 'RTL Isolate',
    '⁨': 'First Strong Isolate',
    '⁩': 'Pop Directional Isolate',
}
_ALL_STEALTH_UNICODE: dict[str, str] = {**_INVISIBLE_UNICODE, **_BIDI_OVERRIDE_UNICODE}


def _find_stealth_chars(text: str) -> tuple[list[str], bool]:
    """Return (unique_char_names_found, any_bidi_override_present)."""
    found: list[str] = []
    seen: set[str] = set()
    any_bidi = False
    for char in text:
        if char in _ALL_STEALTH_UNICODE and char not in seen:
            seen.add(char)
            found.append(_ALL_STEALTH_UNICODE[char])
            if char in _BIDI_OVERRIDE_UNICODE:
                any_bidi = True
    return found, any_bidi


def _render_safe(text: str, max_len: int = 80) -> str:
    """Safe ASCII representation of text for embedding in Finding.detail."""
    parts: list[str] = []
    for char in text[:max_len]:
        if char in _ALL_STEALTH_UNICODE or ord(char) > 127 or ord(char) < 32:
            parts.append(f'\\u{ord(char):04X}')
        else:
            parts.append(char)
    if len(text) > max_len:
        parts.append('...')
    return ''.join(parts)


def check_tool_poisoning(server: MCPServer) -> list[Finding]:
    findings: list[Finding] = []
    full_args = " ".join(server.args)
    # Also build a searchable string from env var VALUES for injection/exfiltration patterns
    full_env_vals = " ".join(f"{k}={v}" for k, v in server.env.items())

    # PI-001: Prompt injection keywords in server args OR env var values
    # Env values: real-world attack — malicious env var instructs server to override system behavior
    for target_text, source_label in [(full_args, "args"), (full_env_vals, "env vars")]:
        for pattern, description in _INJECTION_PATTERNS:
            match = pattern.search(target_text)
            if match:
                findings.append(Finding(
                    check_id="PI-001",
                    title=f"Potential prompt injection in server {source_label} ({description})",
                    detail=(
                        f"Server `{server.name}` contains {source_label} text matching a known prompt injection pattern "
                        f"({description}): `{match.group(0)!r}`. "
                        "This phrasing is used in tool poisoning attacks to silently hijack AI assistant behavior."
                    ),
                    severity=Severity.HIGH,
                    owasp=OWASPCategory.MCP03,
                    server_name=server.name,
                    remediation=(
                        f"Review this server's {source_label} and source code carefully. "
                        "If it was installed from an untrusted source, remove it. "
                        "Legitimate MCP servers do not embed instruction-override language in their configuration."
                    ),
                    engine="custom",
                    cwe_id="CWE-77",
                ))
                break  # One PI-001 per source (args or env)
        else:
            continue
        break  # Found one PI-001 total — don't double-fire

    # PI-002: Excessively long combined args
    if len(full_args) > _MAX_ARGS_LENGTH:
        findings.append(Finding(
            check_id="PI-002",
            title=f"Excessively long server arguments ({len(full_args):,} chars)",
            detail=(
                f"Server `{server.name}` has unusually long combined arguments ({len(full_args):,} characters). "
                f"Arguments exceeding {_MAX_ARGS_LENGTH:,} characters may be hiding injected "
                "instructions or obfuscated payloads within what appears to be normal configuration."
            ),
            severity=Severity.MEDIUM,
            owasp=OWASPCategory.MCP06,
            server_name=server.name,
            remediation=(
                "Review all arguments to this server carefully. "
                "Legitimate MCP servers rarely require more than a few hundred characters of configuration arguments. "
                "Look for any base64-encoded strings or unusually dense text blocks."
            ),
            engine="custom",
            cwe_id="CWE-400",
        ))

    # PI-003: Horizontal scroll hidden injection (MDPI May 2026)
    # Detects the delivery mechanism (single long line), not the payload content
    for i, arg in enumerate(server.args):
        if len(arg) > _HORIZONTAL_SCROLL_THRESHOLD and '\n' not in arg and '\r' not in arg:
            has_injection = any(p.search(arg) for p, _ in _INJECTION_PATTERNS)
            severity = Severity.HIGH if has_injection else Severity.MEDIUM
            findings.append(Finding(
                check_id="PI-003",
                title=f"Horizontal-scroll injection risk: arg #{i+1} is {len(arg):,} chars (single line)",
                detail=(
                    f"Server `{server.name}` has argument #{i+1} with {len(arg):,} characters on a single line. "
                    "In Claude Desktop and Cursor approval dialogs, content beyond the visible viewport "
                    "is hidden via horizontal scroll. An attacker can hide an injected instruction "
                    "off-screen while the visible portion looks harmless. "
                    "(Research source: MDPI May 2026 — MCP Threat Modeling)"
                ),
                severity=severity,
                owasp=OWASPCategory.MCP03,
                server_name=server.name,
                remediation=(
                    "Investigate why this argument is so long. Legitimate MCP server arguments "
                    "are typically short paths, flags, or package names (< 200 chars). "
                    "If you need to pass long configuration, use a config file instead of an inline argument."
                ),
                engine="custom",
                cwe_id="CWE-693",
            ))
            break  # One PI-003 per server

    # PI-004: Obfuscation via escape sequences in args
    # Detects unicode/hex escape sequences used to hide injection payloads beyond what tools display.
    # Real-world attack: "\\u0049\\u0067\\u006e\\u006f\\u0072\\u0065" decodes to "Ignore"
    for i, arg in enumerate(server.args):
        uni_match = _UNICODE_ESC_RE.search(arg)
        hex_match = _HEX_ESC_RE.search(arg)
        obf_match = uni_match or hex_match
        if obf_match:
            enc_type = "unicode escape sequences (\\uXXXX)" if uni_match else "hex escape sequences (\\xXX)"
            findings.append(Finding(
                check_id="PI-004",
                title=f"Obfuscated payload in server args via {enc_type.split('(')[0].strip()}",
                detail=(
                    f"Server `{server.name}` has argument #{i+1} containing {enc_type}: "
                    f"`{obf_match.group(0)!r}`. "
                    "Escape sequences are used in tool poisoning attacks to embed injection payloads "
                    "that look like gibberish in UI displays but decode to instruction-override text "
                    "when interpreted by the language model. This is a strong indicator of adversarial content."
                ),
                severity=Severity.HIGH,
                owasp=OWASPCategory.MCP03,
                server_name=server.name,
                remediation=(
                    f"Remove argument #{i+1} from server `{server.name}` and investigate its source. "
                    "Legitimate MCP server arguments never require escape-sequence-encoded payloads. "
                    "If the server was installed from a third-party source, remove it immediately."
                ),
                engine="custom",
                attack_tactic="defense-evasion",
                cwe_id="CWE-116",
            ))
            break  # One PI-004 per server

    # DX-001: Data exfiltration patterns in args OR env var values
    # Real-world incident (Postmark, Sep 2025): DEFAULT_BCC env var silently exfiltrated all emails
    _dx_sources = [
        (full_args, "args", _EXFILTRATION_PATTERNS),
        (full_env_vals, "env vars", _EXFILTRATION_ENV_PATTERNS),
    ]
    for target_text, source_label, patterns in _dx_sources:
        for pattern, description in patterns:
            match = pattern.search(target_text)
            if match:
                findings.append(Finding(
                    check_id="DX-001",
                    title=f"Potential data exfiltration pattern in server {source_label}: {description}",
                    detail=(
                        f"Server `{server.name}` {source_label} contain language suggesting data exfiltration: "
                        f"`{match.group(0)!r}` ({description}). "
                        "In the Postmark MCP incident (Sep 2025), a malicious env var (DEFAULT_BCC) silently "
                        "BCC'd all emails to an attacker for 2 weeks. "
                        "Legitimate MCP servers do not need data-transfer directives in configuration."
                    ),
                    severity=Severity.HIGH,
                    owasp=OWASPCategory.MCP03,
                    server_name=server.name,
                    remediation=(
                        f"Review this server's {source_label} and source code carefully. "
                        "If the server makes undisclosed outbound connections or email forwarding, remove it. "
                        "Legitimate MCP servers document all network calls and data flows explicitly."
                    ),
                    engine="custom",
                    cwe_id="CWE-200",
                ))
                break
        else:
            continue
        break  # One DX-001 total — don't double-fire

    # PI-005: Invisible / zero-width / bidi-override Unicode steganography
    # Attack: embed U+200B (Zero Width Space) between letters of "ignore all instructions" —
    # the text is invisible in Claude Desktop / Cursor approval dialogs but the LLM reads it
    # intact, bypassing PI-001's keyword regex (which requires consecutive visible characters).
    # Bidi overrides (U+202E) additionally flip display direction (Trojan Source technique).
    _pi005_targets = [
        (" ".join(server.args), "args"),
        (" ".join(f"{k}={v}" for k, v in server.env.items()), "env vars"),
        (server.name, "server name"),
    ]
    for target_text, source_label in _pi005_targets:
        found_names, any_bidi = _find_stealth_chars(target_text)
        if found_names:
            severity = Severity.HIGH if any_bidi else Severity.MEDIUM
            bidi_note = (
                " Bidi override characters (Trojan Source technique, CVE-2021-42574) can "
                "make injected text display as harmless-looking content while the LLM "
                "processes the actual malicious instruction."
            ) if any_bidi else ""
            char_summary = ", ".join(found_names[:3])
            if len(found_names) > 3:
                char_summary += f" (+{len(found_names) - 3} more)"
            findings.append(Finding(
                check_id="PI-005",
                title=f"Invisible Unicode in server {source_label}: {char_summary}",
                detail=(
                    f"Server `{server.name}` {source_label} contain invisible Unicode character(s): "
                    f"{char_summary}. "
                    "These characters render as nothing in Claude Desktop / Cursor approval dialogs "
                    "but are passed to the language model in full, enabling completely invisible "
                    "injection attacks. An attacker can split keywords like 'ignore' into "
                    "'i​g​n​o​r​e' — bypassing PI-001 regex checks while "
                    f"the LLM reads the word normally.{bidi_note}"
                ),
                severity=severity,
                owasp=OWASPCategory.MCP03,
                server_name=server.name,
                remediation=(
                    f"Open this server's {source_label} in a hex editor or Unicode-aware "
                    "text editor to locate and remove the invisible characters. "
                    "These have no legitimate use in MCP server configuration. "
                    "If this config was received from an external source or copy-pasted "
                    "from a chat/web page, treat it as potentially tampered."
                ),
                engine="custom",
                attack_tactic="defense-evasion",
                cwe_id="CWE-116",
            ))
            break  # One PI-005 per server — first field wins to avoid noise

    return findings

File: packages/cli/internal/engine/checks/tool_poisoning.go
package checks

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/abhijitdalal26/MCP-Audit/cli/internal/engine/models"
	"github.com/abhijitdalal26/MCP-Audit/cli/internal/engine/parser"
)

var injectionPatterns = []struct {
	re   *regexp.Regexp
	desc string
}{
	{regexp.MustCompile(`(?i)ignore\s+(previous|all|prior)\s+instructions?`), "instruction override phrase"},
	{regexp.MustCompile(`(?i)do\s+not\s+(reveal|disclose|share|tell)`), "information suppression directive"},
	{regexp.MustCompile(`(?i)(override|bypass|ignore)\s+(the\s+)?(system\s+prompt|instructions?|rules?|constraints?)`), "system prompt override"},
	{regexp.MustCompile(`(?i)you\s+(are|must|should)\s+(now\s+|always\s+)?(act\s+as|pretend|roleplay|simulate)`), "role manipulation directive"},
	{regexp.MustCompile(`(?i)(hidden|secret|confidential)\s+instructions?`), "hidden instruction reference"},
	{regexp.MustCompile(`(?i)<\|.+?\|>`), "control token injection"},
	{regexp.MustCompile(`(?i)\[INST\]|\[/INST\]|<<SYS>>|<</SYS>>`), "LLM control token"},
	{regexp.MustCompile(`(?i)###\s*SYSTEM\s*###|###\s*USER\s*###|###\s*ASSISTANT\s*###`), "LLM section delimiter"},
	{regexp.MustCompile(`(?i)<system>|</system>|<user>|</user>|<assistant>|</assistant>`), "XML role delimiter injection"},
	{regexp.MustCompile(`(?i)AUTOEXECUTE|AUTO_EXECUTE|SYSTEM_OVERRIDE`), "automation override keyword"},
	{regexp.MustCompile(`(?i)disregard\s+(all|any|previous)`), "instruction disregard phrase"},
	{regexp.MustCompile(`(?i)(your|the)\s+(real|true|actual)\s+(purpose|goal|mission|task)\s+is`), "purpose redefinition"},
	{regexp.MustCompile(`(?i)developer\s+mode|jailbreak\s+mode|DAN\s+mode`), "jailbreak mode reference"},
}

var exfilPatterns = []struct {
	re   *regexp.Regexp
	desc string
}{
	{regexp.MustCompile(`(?i)(send|upload|transmit|exfiltrat|forward|relay)\s+(data|files?|credentials?|tokens?)\s+(to|from)`), "data transfer directive"},
	{regexp.MustCompile(`(?i)POST\s+to\s+https?://`), "HTTP POST to external URL"},
	{regexp.MustCompile(`https?://(?:(?:localhost|127\.0\.0\.1|0\.0\.0\.0)[^\s]*|[^\s]{15,})`), "external URL embedded in argument"},
	{regexp.MustCompile(`(?i)(webhook|callback)\s+url`), "webhook/callback URL reference"},
	{regexp.MustCompile(`(?i)(steal|harvest|collect|scrape)\s+(data|credentials?|tokens?|passwords?)`), "data harvesting language"},
	{regexp.MustCompile(`(?i)(bcc|blind.?carbon.?copy)`), "BCC/blind copy email exfiltration"},
	{regexp.MustCompile(`(?i)(forward.?to|cc.?to|reply.?to)\s*[=:]\s*[^\s@]+@[^\s]{3,}`), "email forwarding rule"},
}

// exfilEnvPatterns: subset of exfilPatterns safe for env var values (excludes generic URL pattern).
var exfilEnvPatterns = []struct {
	re   *regexp.Regexp
	desc string
}{
	{regexp.MustCompile(`(?i)(send|upload|transmit|exfiltrat|forward|relay)\s+(data|files?|credentials?|tokens?)\s+(to|from)`), "data transfer directive"},
	{regexp.MustCompile(`(?i)POST\s+to\s+https?://`), "HTTP POST to external URL"},
	{regexp.MustCompile(`(?i)(webhook|callback)\s+url`), "webhook/callback URL reference"},
	{regexp.MustCompile(`(?i)(steal|harvest|collect|scrape)\s+(data|credentials?|tokens?|passwords?)`), "data harvesting language"},
	{regexp.MustCompile(`(?i)(bcc|blind.?carbon.?copy)`), "BCC/blind copy email exfiltration"},
	{regexp.MustCompile(`(?i)(forward.?to|cc.?to|reply.?to)\s*[=:]\s*[^\s@]+@[^\s]{3,}`), "email forwarding rule"},
}

var unicodeEscRE = regexp.MustCompile(`(?:\\u[0-9a-fA-F]{4}){4,}`)
var hexEscRE = regexp.MustCompile(`(?:\\x[0-9a-fA-F]{2}){4,}`)

// invisibleUnicode: zero-width and invisible Unicode codepoints, keyed by rune value.
// Mirrors Python's _INVISIBLE_UNICODE + _BIDI_OVERRIDE_UNICODE.
// Codepoints are expressed as integer literals to avoid embedding invisible chars in source.
var invisibleUnicode = map[rune]string{
	0x200B: "Zero Width Space",
	0x200C: "Zero Width Non-Joiner",
	0x200D: "Zero Width Joiner",
	0xFEFF: "BOM/Zero Width No-Break Space",
	0x2060: "Word Joiner",
	0x2061: "Function Application (invisible)",
	0x2062: "Invisible Times",
	0x2063: "Invisible Separator",
	0x2064: "Invisible Plus",
	0x00AD: "Soft Hyphen",
	0x180E: "Mongolian Vowel Separator",
	0x034F: "Combining Grapheme Joiner",
}

var bidiOverrideUnicode = map[rune]string{
	0x202A: "LTR Embedding",
	0x202B: "RTL Embedding",
	0x202C: "Pop Directional Formatting",
	0x202D: "LTR Override",
	0x202E: "RTL Override (Trojan Source)",
	0x2066: "LTR Isolate",
	0x2067: "RTL Isolate",
	0x2068: "First Strong Isolate",
	0x2069: "Pop Directional Isolate",
}

const maxArgsLength = 2000
const horizontalScrollThreshold = 300

func findStealthChars(text string) (names []string, anyBidi bool) {
	seen := map[rune]bool{}
	for _, r := range text {
		if seen[r] {
			continue
		}
		if name, ok := invisibleUnicode[r]; ok {
			seen[r] = true
			names = append(names, name)
		} else if name, ok := bidiOverrideUnicode[r]; ok {
			seen[r] = true
			names = append(names, name)
			anyBidi = true
		}
	}
	return
}

// CheckToolPoisoning checks for prompt injection, exfiltration, and Unicode steganography (PI-001..005, DX-001).
func CheckToolPoisoning(server *parser.MCPServer) []models.Finding {
	findings := []models.Finding{}

	fullArgs := strings.Join(server.Args, " ")
	var envParts []string
	for k, v := range server.Env {
		envParts = append(envParts, k+"="+v)
	}
	fullEnvVals := strings.Join(envParts, " ")

	// PI-001: Prompt injection keywords in args or env var values
	pi001Done := false
	for _, src := range []struct{ text, label string }{{fullArgs, "args"}, {fullEnvVals, "env vars"}} {
		if pi001Done {
			break
		}
		for _, pat := range injectionPatterns {
			if m := pat.re.FindString(src.text); m != "" {
				findings = append(findings, models.Finding{
					CheckID:  "PI-001",
					Title:    fmt.Sprintf("Potential prompt injection in server %s (%s)", src.label, pat.desc),
					Detail:   fmt.Sprintf("Server `%s` contains %s text matching a known prompt injection pattern (%s): `%s`. This phrasing is used in tool poisoning attacks to silently hijack AI assistant behavior.", server.Name, src.label, pat.desc, m),
					Severity: models.SeverityHigh,
					OWASP:    models.MCP03,
					ServerName: server.Name,
					Remediation: fmt.Sprintf("Review this server's %s and source code carefully. If it was installed from an untrusted source, remove it. Legitimate MCP servers do not embed instruction-override language in their configuration.", src.label),
					Engine:   "custom",
					CWEID:    "CWE-77",
				})
				pi001Done = true
				break
			}
		}
	}

	// PI-002: Excessively long combined args
	if len(fullArgs) > maxArgsLength {
		findings = append(findings, models.Finding{
			CheckID:  "PI-002",
			Title:    fmt.Sprintf("Excessively long server arguments (%d chars)", len(fullArgs)),
			Detail:   fmt.Sprintf("Server `%s` has unusually long combined arguments (%d characters). Arguments exceeding %d characters may be hiding injected instructions or obfuscated payloads within what appears to be normal configuration.", server.Name, len(fullArgs), maxArgsLength),
			Severity: models.SeverityMedium,
			OWASP:    models.MCP06,
			ServerName: server.Name,
			Remediation: "Review all arguments to this server carefully. Legitimate MCP servers rarely require more than a few hundred characters of configuration arguments. Look for any base64-encoded strings or unusually dense text blocks.",
			Engine:   "custom",
			CWEID:    "CWE-400",
		})
	}

	// PI-003: Horizontal scroll hidden injection
	for i, arg := range server.Args {
		if len(arg) > horizontalScrollThreshold && !strings.ContainsAny(arg, "\n\r") {
			hasInj := false
			for _, pat := range injectionPatterns {
				if pat.re.MatchString(arg) {
					hasInj = true
					break
				}
			}
			sev := models.SeverityMedium
			if hasInj {
				sev = models.SeverityHigh
			}
			findings = append(findings, models.Finding{
				CheckID:  "PI-003",
				Title:    fmt.Sprintf("Horizontal-scroll injection risk: arg #%d is %d chars (single line)", i+1, len(arg)),
				Detail:   fmt.Sprintf("Server `%s` has argument #%d with %d characters on a single line. In Claude Desktop and Cursor approval dialogs, content beyond the visible viewport is hidden via horizontal scroll. An attacker can hide an injected instruction off-screen while the visible portion looks harmless. (Research source: MDPI May 2026 — MCP Threat Modeling)", server.Name, i+1, len(arg)),
				Severity: sev,
				OWASP:    models.MCP03,
				ServerName: server.Name,
				Remediation: "Investigate why this argument is so long. Legitimate MCP server arguments are typically short paths, flags, or package names (< 200 chars). If you need to pass long configuration, use a config file instead of an inline argument.",
				Engine:   "custom",
				CWEID:    "CWE-693",
			})
			break
		}
	}

	// PI-004: Obfuscation via escape sequences in args
	for i, arg := range server.Args {
		uniMatch := unicodeEscRE.FindString(arg)
		hexMatch := hexEscRE.FindString(arg)
		if uniMatch != "" || hexMatch != "" {
			encType := "unicode escape sequences (\\uXXXX)"
			match := uniMatch
			if uniMatch == "" {
				encType = "hex escape sequences (\\xXX)"
				match = hexMatch
			}
			findings = append(findings, models.Finding{
				CheckID:  "PI-004",
				Title:    fmt.Sprintf("Obfuscated payload in server args via %s", strings.SplitN(encType, "(", 2)[0]),
				Detail:   fmt.Sprintf("Server `%s` has argument #%d containing %s: `%s`. Escape sequences are used in tool poisoning attacks to embed injection payloads that look like gibberish in UI displays but decode to instruction-override text when interpreted by the language model. This is a strong indicator of adversarial content.", server.Name, i+1, encType, match),
				Severity: models.SeverityHigh,
				OWASP:    models.MCP03,
				ServerName: server.Name,
				Remediation: fmt.Sprintf("Remove argument #%d from server `%s` and investigate its source. Legitimate MCP server arguments never require escape-sequence-encoded payloads. If the server was installed from a third-party source, remove it immediately.", i+1, server.Name),
				Engine:       "custom",
				AttackTactic: "defense-evasion",
				CWEID:        "CWE-116",
			})
			break
		}
	}

	// DX-001: Data exfiltration patterns in args or env var values
	dx001Done := false
	for _, src := range []struct {
		text     string
		label    string
		patterns []struct {
			re   *regexp.Regexp
			desc string
		}
	}{
		{fullArgs, "args", exfilPatterns},
		{fullEnvVals, "env vars", exfilEnvPatterns},
	} {
		if dx001Done {
			break
		}
		for _, pat := range src.patterns {
			if m := pat.re.FindString(src.text); m != "" {
				findings = append(findings, models.Finding{
					CheckID:  "DX-001",
					Title:    fmt.Sprintf("Potential data exfiltration pattern in server %s: %s", src.label, pat.desc),
					Detail:   fmt.Sprintf("Server `%s` %s contain language suggesting data exfiltration: `%s` (%s). In the Postmark MCP incident (Sep 2025), a malicious env var (DEFAULT_BCC) silently BCC'd all emails to an attacker for 2 weeks. Legitimate MCP servers do not need data-transfer directives in configuration.", server.Name, src.label, m, pat.desc),
					Severity: models.SeverityHigh,
					OWASP:    models.MCP03,
					ServerName: server.Name,
					Remediation: fmt.Sprintf("Review this server's %s and source code carefully. If the server makes undisclosed outbound connections or email forwarding, remove it. Legitimate MCP servers document all network calls and data flows explicitly.", src.label),
					Engine:   "custom",
					CWEID:    "CWE-200",
				})
				dx001Done = true
				break
			}
		}
	}

	// PI-005: Invisible / zero-width / bidi-override Unicode steganography
	pi005Sources := []struct{ text, label string }{
		{fullArgs, "args"},
		{fullEnvVals, "env vars"},
		{server.Name, "server name"},
	}
	for _, src := range pi005Sources {
		names, anyBidi := findStealthChars(src.text)
		if len(names) == 0 {
			continue
		}
		sev := models.SeverityMedium
		if anyBidi {
			sev = models.SeverityHigh
		}
		summary := strings.Join(names, ", ")
		if len(names) > 3 {
			summary = strings.Join(names[:3], ", ") + fmt.Sprintf(" (+%d more)", len(names)-3)
		}
		bidiNote := ""
		if anyBidi {
			bidiNote = " Bidi override characters (Trojan Source technique, CVE-2021-42574) can make injected text display as harmless-looking content while the LLM processes the actual malicious instruction."
		}
		findings = append(findings, models.Finding{
			CheckID:  "PI-005",
			Title:    fmt.Sprintf("Invisible Unicode in server %s: %s", src.label, summary),
			Detail:   fmt.Sprintf("Server `%s` %s contain invisible Unicode character(s): %s. These characters render as nothing in Claude Desktop / Cursor approval dialogs but are passed to the language model in full, enabling completely invisible injection attacks. An attacker can split keywords like 'ignore' into separated characters — bypassing PI-001 regex checks while the LLM reads the word normally.%s", server.Name, src.label, summary, bidiNote),
			Severity: sev,
			OWASP:    models.MCP03,
			ServerName: server.Name,
			Remediation: fmt.Sprintf("Open this server's %s in a hex editor or Unicode-aware text editor to locate and remove the invisible characters. These have no legitimate use in MCP server configuration. If this config was received from an external source or copy-pasted from a chat/web page, treat it as potentially tampered.", src.label),
			Engine:       "custom",
			AttackTactic: "defense-evasion",
			CWEID:        "CWE-116",
		})
		break // one PI-005 per server
	}

	return findings
}
